aptos-move/move-examples/data_structures/sources/smart_tree.move

```diff
@@ -0,0 +1,971 @@
+    /// Requires the tree is not changed after the input iterator is generated.
+    public fun next_iter_or_die<V>(tree: &SmartTree<V>, iter: Iterator): Iterator {
+        let node_index = iter.node_index;
+
+        let node = bucket_table::borrow(&tree.nodes, node_index);
+        iter.child_index = iter.child_index + 1;
+        if (iter.child_index < vector::length(&node.children)) {
+            iter.key = vector::borrow(&node.children, iter.child_index).max_key;
+            return iter
+        };
+
+        let next_index = node.next;
+        if (next_index != NULL_INDEX) {
+            let next_node = bucket_table::borrow(&tree.nodes, next_index);
+            iter.node_index = next_index;
+            iter.child_index = 0;
+            iter.key = vector::borrow(&next_node.children, 0).max_key;
+            return iter
+        };
+
+        new_end_iter(tree)
+    }
+
+    /// Returns the previous iterator, or none if already at the begin iterator.
+    /// Requires the tree is not changed after the input iterator is generated.
+    public fun prev_iter<V>(tree: &SmartTree<V>, iter: Iterator): Option<Iterator> {
+        if (iter.node_index == tree.min_leaf_index && iter.child_index == 0) {
+            return option::none()
+        };
+
+        option::some(prev_iter_or_die(tree, iter))
+    }
+
+    /// Returns the previous iterator, aborts if already at the begin iterator.
+    /// Requires the tree is not changed after the input iterator is generated.
+    public fun prev_iter_or_die<V>(tree: &SmartTree<V>, iter: Iterator): Iterator {
+        let node_index = iter.node_index;
+
+        let prev_index;
+
+        if (node_index == NULL_INDEX) {
+            prev_index = tree.max_leaf_index;
+        } else {
+            let node = bucket_table::borrow(&tree.nodes, node_index);
+            if (iter.child_index >= 1) {
+                iter.child_index = iter.child_index - 1;
+                iter.key = vector::borrow(&node.children, iter.child_index).max_key;
+                return iter
+            };
+            prev_index = node.prev;
+        };
+
+        assert!(prev_index != NULL_INDEX, E_INTERNAL);
+
+        let prev_node = bucket_table::borrow(&tree.nodes, prev_index);
+        let len = vector::length(&prev_node.children);
+        iter.node_index = prev_index;
+        iter.child_index = len - 1;
+        iter.key = vector::borrow(&prev_node.children, len - 1).max_key;
+        iter
+    }
+
+    //////////////////////////////
+    // Internal Implementations //
+    //////////////////////////////
+
+    fun new_node(is_leaf: bool, parent: u64): Node {
+        Node {
+            is_leaf: is_leaf,
+            parent: parent,
+            children: vector::empty(),
+            prev: NULL_INDEX,
+            next: NULL_INDEX,
+        }
+    }
+
+    fun new_child(max_key: u64, index: u64): Child {
+        Child {
+            max_key: max_key,
+            index: index,
```

Comment:
I think this line should be `node_index: index,`. By the way, I am wondering if this compiled ok?

Reply:
I was in the process of renaming some variables to improve readability, mistakenly pushed a bad version, fixed now.